### ISSUE TITLE
CA2225: Don't ask for cast alternates

### ISSUE THREAD

**Issue by dcwuser** [dotnet/roslyn-analyzers#1706]
Write a library that defines an implicit or explicit cast, e.g.

        public static implicit operator Int128 (Int64 value) {
            // create a 128-bit integer from a 64-bit integer
        }

Run code analysis against the library. It will fire a CA2225 violation that says "Provide a method named 'ToInt128' or 'FromInt128' as an alternate for operator op_implicit."

This is, I believe, new to the Roslyn implementation. FxCop asked for alternates for a long list of operator overloads (https://msdn.microsoft.com/en-us/library/ms182355.aspx), but not for cast operators.

Given that none of the core types of the BCL even begin to implement this rule (e.g. there are no named methods for casting between int, long, double, float, etc.), this seems a ridiculous requirement for library authors.

**Comment by Echo-8-ERA** [comment]
Somewhat related is that the alternatives it looks for don't make sense.

If I have
```c#
class foo {
public foo(int a) { … }
public static explicit operator foo(int a) => new foo(a);
}
```
I would have expected it to look for `FromInt` as opposed to `FromFoo`.

**Comment by sharwell** [comment]
@dcwuser In the documentation you linked to is the following note after the long table:

> The rule also checks implicit and explicit cast operators in a type (`SomeType`) by checking for methods named `ToSomeType` and `FromSomeType`

**Comment by sharwell** [comment]
@Echo-8-ERA Yes. Please file a separate issue for that since it's an obvious oversight/bug.

**Comment by Echo-8-ERA** [comment]
@sharwell Done (#1717)

**Comment by dcwuser** [comment]
@sharwell Granted, I missed that, and thanks for pointing it out. (Why wouldn't it be in the table?) Still I stand by my request with the remaining arguments:

1. FxCop in VS2015 didn't fire on this. So 1000s of compliant libraries have been produced and now you are demanding that they change their public surface area.

2. More importantly, unlike for the other named alternates, the BCL itself doesn't obey this. So you are asking us all to change our public surface areas in a way that doesn't conform to BCL design.

**Comment by mavasani** [comment]
> FxCop in VS2015 didn't fire on this. 

Can you please give a concrete source example? I tried the following source code, and binary FxCop does generate a CA2225:

```
using System;

public class MyInt
{
    public static implicit operator MyInt(Int64 value)
    {
        return 0;
    }
}
```

**Comment by mavasani** [comment]
Closing this as by-design, current implementation matches binary FxCop (at least for the repros available in this issue) and also matches documentation.
